compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -6291,26 +6293,37 @@ private BType getCompletionType(BType collectionType, BLangQueryExpr queryExpr,
         return completionType;
     }
 
-    private BType getNonContextualQueryType(BType staticType, BType basicType) {
-        BType resultType;
+    private BType getNonContextualQueryType(BType staticType, BType basicType, Location pos) {
         switch (basicType.tag) {
             case TypeTags.TABLE:
-                resultType = symTable.tableType;
+                if (types.isAssignable(staticType, symTable.mapAllType)) {
+                    return symTable.tableType;
+                }
                 break;
             case TypeTags.STREAM:
-                resultType = symTable.streamType;
-                break;
+                return symTable.streamType;
+            case TypeTags.MAP:
+                dlog.error(pos, INCOMPATIBLE_QUERY_CONSTRUCT_MAP_TYPE);
+                return symTable.semanticError;
             case TypeTags.XML:
-                resultType = new BXMLType(staticType, null);
+                if (types.isSubTypeOfBaseType(staticType, symTable.xmlType.tag)) {
+                    return new BXMLType(staticType, null);
+                }
                 break;
             case TypeTags.STRING:
-                resultType = symTable.stringType;
+                if (types.isSubTypeOfBaseType(staticType, TypeTags.STRING)) {
+                    return symTable.stringType;
+                }
                 break;
+            case TypeTags.ARRAY:
+            case TypeTags.TUPLE:
+            case TypeTags.OBJECT:
+                return new BArrayType(staticType);
             default:
-                resultType = new BArrayType(staticType);
-                break;
+                return symTable.semanticError;
         }
-        return resultType;
+        dlog.error(pos, INCOMPATIBLE_QUERY_CONSTRUCT_TYPE, staticType, basicType);
```

Comment:
For something like this
```
    table<Employee> key(name) t = table [
        { name: "John", salary: 100 },
        { name: "Jane", salary: 200 }
    ];

    var res = from Employee i in t
    select 1;
```
the error would be `incompatible type 'int', expected a construct type of 'table<Employee> key(name)'`. Shouldn't the error be something like
`incompatible types: 'table' cannot be constrained with 'int'`, since the basic type that the query can construct is a 'table' here.

Reply:
We might need to consider the referred type here to handle the type reference types.
Now the following code crashes at runtime
```ballerina
type T1 table<Employee> key(name);

public function main() {
    T1 t = table [
        { name: "John", salary: 100 },
        { name: "Jane", salary: 200 }
    ];

    var _ = from var i in t
    select 1;
}
```

Reply:
Updated.